usb/imx-usbc/client.c

```diff
@@ -0,0 +1,270 @@
+
+		/* Low frequency interrupts, handle for OUT control endpoint */
+		if ((imx_common.dc.setupstat & 0x1) && imx_common.dc.runIrqThread)
+			desc_classSetup(&imx_common.dc.setup);
+
+		ctrl_lfIrq();
+	}
+	mutexUnlock(imx_common.dc.irqLock);
+
+	endthread();
+}
+
+
+static void usbclient_cleanData(void)
+{
+	int i;
+
+	usbclient_buffDestory((void *)imx_common.dc.base, USB_BUFFER_SIZE);
+	usbclient_buffDestory((void *)imx_common.data.setupMem, USB_BUFFER_SIZE);
+
+	usbclient_buffDestory((void *)imx_common.dc.dtdMem, USB_BUFFER_SIZE);
+	usbclient_buffDestory((void *)imx_common.dc.endptqh, USB_BUFFER_SIZE);
+
+	for (i = 0; i < ENDPOINTS_NUMBER; ++i) {
+		if (imx_common.data.endpts[i].caps[USB_ENDPT_DIR_IN].init) {
+			usbclient_buffDestory((void *)imx_common.data.endpts[i].buf[USB_ENDPT_DIR_IN].vBuffer, USB_BUFFER_SIZE);
+			imx_common.data.endpts[i].buf[USB_ENDPT_DIR_IN].vBuffer = NULL;
+		}
+
+		if (imx_common.data.endpts[i].caps[USB_ENDPT_DIR_OUT].init) {
+			usbclient_buffDestory((void *)imx_common.data.endpts[i].buf[USB_ENDPT_DIR_OUT].vBuffer, USB_BUFFER_SIZE);
+			imx_common.data.endpts[i].buf[USB_ENDPT_DIR_OUT].vBuffer = NULL;
+		}
+	}
+}
+
+
+int usbclient_init(usb_desc_list_t *desList)
+{
+	int i;
+	int res = 0;
+
+	imx_common.dc.irqLock = 0;
+	imx_common.dc.irqCond = 0;
+	imx_common.dc.endp0Lock = 0;
+	imx_common.dc.endp0Cond = 0;
+	imx_common.dc.runIrqThread = 1;
+
+	imx_common.dc.dev_addr = 0;
+	imx_common.dc.base = phy_getBase(USB_BUFFER_SIZE);
+
+	if (imx_common.dc.base == MAP_FAILED)
+		return -ENOMEM;
+
+	imx_common.data.setupMem = usbclient_allocBuff(USB_BUFFER_SIZE);
+
+	if (imx_common.data.setupMem == MAP_FAILED){
+		usbclient_cleanData();
+		return -ENOMEM;
+	}
+
+	for (i = 1; i < ENDPOINTS_NUMBER; ++i) {
+		imx_common.data.endpts[i].caps[USB_ENDPT_DIR_OUT].init = 0;
+		imx_common.data.endpts[i].caps[USB_ENDPT_DIR_IN].init = 0;
+	}
+
+	if (desc_init(desList, &imx_common.data, &imx_common.dc) < 0) {
+		usbclient_cleanData();
+		return -ENOMEM;
+	}
+
+	if (mutexCreate(&imx_common.dc.irqLock) != EOK) {
+		usbclient_cleanData();
+		return -ENOENT;
+	}
+
+	if (condCreate(&imx_common.dc.irqCond) != EOK) {
+		usbclient_cleanData();
+		resourceDestroy(imx_common.dc.irqLock);
+		return -ENOENT;
+	}
+
+	if (mutexCreate(&imx_common.dc.endp0Lock) != EOK) {
+		usbclient_cleanData();
+		resourceDestroy(imx_common.dc.irqLock);
+		resourceDestroy(imx_common.dc.irqCond);
+		return -ENOENT;
+	}
+
+	if (condCreate(&imx_common.dc.endp0Cond) != EOK) {
+		usbclient_cleanData();
+		resourceDestroy(imx_common.dc.irqLock);
+		resourceDestroy(imx_common.dc.irqCond);
+		resourceDestroy(imx_common.dc.endp0Lock);
+		return -ENOENT;
+	}
+
+	if (ctrl_init(&imx_common.data, &imx_common.dc) < 0) {
+		usbclient_cleanData();
+		resourceDestroy(imx_common.dc.irqLock);
+		resourceDestroy(imx_common.dc.irqCond);
+		resourceDestroy(imx_common.dc.endp0Lock);
+		resourceDestroy(imx_common.dc.endp0Cond);
+		return -ENOENT;
+	}
+
```

Comment:
code duplication could be avoided. we can obviously use goto but we may not do it. stil i think 
`int res = EOK;
[...]
res |= mutexCreate(&imx_common.dc.endp0Lock);
res | = condCreate(&imx_common.dc.endp0Cond);
[...]
if (res != EOK) {
		resourceDestroy(imx_common.dc.irqLock);
 		resourceDestroy(imx_common.dc.irqCond);
               [...]
}`
 would be a little bit cleaner.
By no means any code should be changed it is just general idea.